emails/views.py

```diff
@@ -271,50 +274,40 @@ def _sns_message(message_json):
         incr_if_enabled('email_for_noreply_address', 1)
         return HttpResponse('noreply address is not supported.')
 
-    local_portion_hash = sha256(local_portion.encode('utf-8')).hexdigest()
+    domain_portion = to_address.split('@')[1]
 
     try:
-        relay_address = RelayAddress.objects.get(address=local_portion)
-    except RelayAddress.DoesNotExist:
-        try:
-            DeletedAddress.objects.get(
-                address_hash=local_portion_hash
-            )
-            incr_if_enabled('email_for_deleted_address', 1)
-            # TODO: create a hard bounce receipt rule in SES
-        except DeletedAddress.DoesNotExist:
-            incr_if_enabled('email_for_unknown_address', 1)
-            logger.error(
-                'Received email for unknown address.',
-                extra={'to_address': to_address}
-            )
-        except DeletedAddress.MultipleObjectsReturned:
-            # not sure why this happens on stage but let's handle it
-            incr_if_enabled('email_for_deleted_address_multiple', 1)
+        # FIXME: this ambiguous return of either
+        # RelayAddress or DomainAddress types makes the Rustacean in me throw
+        # up a bit.
```

Comment:
Since I am making the changes let's discuss this. According to [this documentation](https://docs.djangoproject.com/en/3.1/topics/db/models/#model-inheritance) about Django Model Inheritance it seems like there are 2 options:

1. Create an abstract base class called `Alias` and we can move the following properties:
```
    num_forwarded = models.PositiveSmallIntegerField(default=0)
    num_blocked = models.PositiveSmallIntegerField(default=0)
    num_spam = models.PositiveSmallIntegerField(default=0)
```
to `Alias` and have `RelayAddress`, `DeletedAddress`, and `DomainAddress` inherit from `Alias`

 - Pro: we can make sure that the return value is of type `Alias`.
 - Con: we would be editing existing models that already have real data and refactoring this can get tricky

2. Make `DomainAddress` inherit from `RelayAddress` and have separate tables with [multi-table inheritance](https://docs.djangoproject.com/en/3.1/topics/db/models/#multi-table-inheritance).
 

- Pro: we can reuse the `delete` for `DomainAddress` thus reducing duplicate code while being able to reuse logic for handling deleted aliases. Also in this call, we can check that return type is of `RelayAddress`.
- Con: RelayAddress table will get big since any new entry to `DomainAddress` results into a entry in the `RelayAddress` table due to OneToOneField mapping between the child model and the parent.

3. Keep the two models separate.

- Pro: It's clear that there are two separate models and we do not have the complications/cons mentioned in (1) and (2)
- Con: We may have duplicate code

What do you think is the best way forward and why?

Reply:
Hmm ... I think I like option 3 with other techniques to reduce code duplication - e.g., utility functions, or composition over inheritance.